Source/DafnyCore/Compilers/Dafny/DAM.dfy

```diff
@@ -0,0 +1,654 @@
+
+      case Select(record, lbl) => (
+        match SynthStmt(g, record)
+        case Some(Record(fields)) =>
+          mapGet(fields, lbl)
+        case _ => None
+      )
+
+      case Rec(self, fix, body) =>
+        if CheckStmt(g[self := Pos.Thunk(fix)], body, fix) then
+          Some(fix)
+        else None
+
+      case New(init, var_, next) =>
+        var t :- SynthExpr(g, init);
+        SynthStmt(g[var_ := Pos.Ref(t)], next)
+
+      case Read(ref, var_, next) => (
+        match SynthExpr(g, ref)
+        case Some(Ref(t)) => SynthStmt(g[var_ := t], next)
+        case _            => None
+      )
+
+      case Write(lhs, rhs, next) => (
+        match SynthExpr(g, lhs)
+        case Some(Ref(t)) =>
+          if CheckExpr(g, rhs, t) then
+            SynthStmt(g, next)
+          else
+            None
+        case _            => None
+      )
+
+      case _ => None
+    }
+
+    // G |- stmt <= t-
+    predicate CheckStmt(g: Context, stmt: Stmt, t: Neg) decreases stmt, 1 {
+      SynthStmt(g, stmt) == Some(t)
+    }
+
+    // S |- env => G
+    function SynthEnv(s: StoreTyping, env: Env): Option<Context> {
+      mapOption(map var_ <- env :: SynthVal(s, env[var_]))
+    }
+
+    // S |- env <= G
+    predicate CheckEnv(s: StoreTyping, env: Env, g: Context) {
+      SynthEnv(s, env) == Some(g)
+    }
+
+    // S |- (env, expr) => t+
+    function SynthClosedExpr(s: StoreTyping, expr: ClosedExpr): Option<Pos> {
+      var (env, expr) := expr;
+      var g :- SynthEnv(s, env);
+      SynthExpr(g, expr)
+    }
+
+    // S |- (env, expr) <= t+
+    predicate CheckClosedExpr(s: StoreTyping, expr: ClosedExpr, t: Pos) {
+      SynthClosedExpr(s, expr) == Some(t)
+    }
+
+    // S |- (env, stmt) => t-
+    function SynthClosedStmt(s: StoreTyping, stmt: ClosedStmt): Option<Neg> {
+      var (env, stmt) := stmt;
+      var g :- SynthEnv(s, env);
+      SynthStmt(g, stmt)
+    }
+
+    // S |- start- => stack => end-
+    function SynthStack(s: StoreTyping, start: Neg, stack: Stack): Option<Neg> decreases stack {
+      match stack
+      case Empty => Some(start)
+      case Push(Bind(var_, (env, rhs)), stack) => (
+        match start
+        case Value(t) =>
+          var g :- SynthEnv(s, env);
+          var start :- SynthStmt(g[var_ := t], rhs);
+          SynthStack(s, start, stack)
+        case _        => None
+      )
+
+      case Push(Select(lbl), stack) => (
+        match start
+        case Record(fields) =>
+          var start :- mapGet(fields, lbl);
+          SynthStack(s, start, stack)
+        case _ => None
+      )
+
+      case Push(Call(arg), stack) => (
+        match start
+        case Function(dom, cod) =>
+          if CheckClosedExpr(s, arg, dom) then
+            SynthStack(s, cod, stack)
+          else
+            None
+        case _ => None
+      )
+    }
+
+    // S |- stack <= start- <= end-
+    predicate CheckStack(s: StoreTyping, start: Neg, stack: Stack, end: Neg) {
```

Comment:
The name `CheckStack` sounds operational, using the verb "check". Wouldn't it be more appropriate to think of a predicate is using an adjective, like `ValidStack`?

For that matter, it is also possible to rename the predicate just `Valid` and declare it as a member of the `StoreTyping` type.

Reply:
+1 for grouping checking/synthesis functions into each type definition

The terminology Check/Synth is a reference to the two modes of bidirectional typing (checking against a type vs. synthesis of a type).